src/analyzer/protocol/ssl/spicy/SSL.spicy

```diff
@@ -0,0 +1,1901 @@
+    TLSCiphers::TLS_DHE_PSK_WITH_CAMELLIA_256_CBC_SHA384,
+    TLSCiphers::TLS_DHE_RSA_WITH_AES_128_CCM,
+    TLSCiphers::TLS_DHE_RSA_WITH_AES_256_CCM,
+    TLSCiphers::TLS_DHE_RSA_WITH_AES_128_CCM_8,
+    TLSCiphers::TLS_DHE_RSA_WITH_AES_256_CCM_8,
+    TLSCiphers::TLS_DHE_PSK_WITH_AES_128_CCM,
+    TLSCiphers::TLS_DHE_PSK_WITH_AES_256_CCM,
+    TLSCiphers::TLS_PSK_DHE_WITH_AES_128_CCM_8,
+    TLSCiphers::TLS_PSK_DHE_WITH_AES_256_CCM_8,
+    TLSCiphers::TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256_OLD,
+    TLSCiphers::TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+    TLSCiphers::TLS_DHE_PSK_WITH_CHACHA20_POLY1305_SHA256,
+    # DH-anon suites
+    TLSCiphers::TLS_DH_ANON_EXPORT_WITH_RC4_40_MD5,
+    TLSCiphers::TLS_DH_ANON_WITH_RC4_128_MD5,
+    TLSCiphers::TLS_DH_ANON_EXPORT_WITH_DES40_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_DES_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_3DES_EDE_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_AES_128_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_AES_256_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_CAMELLIA_128_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_AES_128_CBC_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_AES_256_CBC_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_CAMELLIA_256_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_SEED_CBC_SHA,
+    TLSCiphers::TLS_DH_ANON_WITH_AES_128_GCM_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_AES_256_GCM_SHA384,
+    TLSCiphers::TLS_DH_ANON_WITH_CAMELLIA_128_CBC_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_CAMELLIA_256_CBC_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_ARIA_128_CBC_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_ARIA_256_CBC_SHA384,
+    TLSCiphers::TLS_DH_ANON_WITH_ARIA_128_GCM_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_ARIA_256_GCM_SHA384,
+    TLSCiphers::TLS_DH_ANON_WITH_CAMELLIA_128_GCM_SHA256,
+    TLSCiphers::TLS_DH_ANON_WITH_CAMELLIA_256_GCM_SHA384
+      -> dh_server_key_exchange : DhClientKeyExchange(len);
+
+    *
+      -> key : bytes &size=len;
+  };
+};
+
+type RsaClientKeyExchange = unit(len: uint64) {
+  rsa_pms : bytes &size=len;
+};
+
+type DhClientKeyExchange= unit(len: uint64) {
+  dh_Yc : bytes &size=len;
+};
+
+type EcdhClientKeyExchange = unit(len: uint64) {
+  point : bytes &size=len;
+};
+
+
+type NewSessionTicket = unit {
+  ticket_lifetime_hint: uint32;
+  ticket_length: uint16;
+  ticket: bytes &size=self.ticket_length;
+};
+
+type Certificate = unit(inout sh: Share) {
+  length: bytes &size=3 &convert=$$.to_uint(spicy::ByteOrder::Network);
+  certificate_list: SingleCertificate[] &size=self.length;
+};
+
+type SingleCertificate = unit {
+  length: bytes &size=3 &convert=$$.to_uint(spicy::ByteOrder::Network);
+  cert: bytes &size=self.length; # certificates, forward to whatever
+};
+
+
+# Conversion and Zeek related functions
+
+import zeek;
+
+public function convert_server_names(snl: SSL::ServerNameList) : vector<bytes> {
+  local out: vector<bytes>;
+
+  for ( i in snl.server_name_list )
+    out.push_back(i.host_name);
+
+  return out;
+}
```

Comment:
My gut reaction to seeing all these conversion functions returning lists is wondering whether we couldn't instead directly transform e.g., `ServerNameList` values to `vector<bytes>` on parsing with a `&convert`. That way we would perform this conversion exactly once.

Here and below.

Reply:
Just for my understanding - is the conversion currently performed more than once?

My understanding is that currently this is performed when the event `ssl_extension_server_name` is raised (via the evt file). Which is raised once, and should perform this conversion once.

Same for the others - I don't think that the conversions are performed repeatedly.

Reply:
Ah okay, I didn't fully trace the code. This was more a comment on style from reviewing internal code which called potentially expensive conversion functions repeatedly.

I'd say if something _always_ needs conversion (even if only once) my gut reaction would be to do that during parsing; `&convert` is exactly the tool to change from "on wire representation for parsing" to "user-consumable representation". One benefit is that should one suddenly expose something in more places the conversion already happened and one does not pay for it again; another benefit is that parsed datum and conversion are close together (e.g., here we could probably change `convert_server_names` to work on `ServerNameList::server_name_list` instead of the full unit and then `&convert` just that field (or even strip away `length` by adding a `&convert` to the unit.

If we sometimes need conversion and sometimes don't the situation is more tricky, but I'd probably still cache it in a unit var which I'd populate from a field hook as soon as the data becomes available.

This is basically trying to balance "accidentally converting repeatedly" and "computing a value which is not needed" (e.g., if there is no EVT hook mapping to expose a computed value).